### ISSUE TITLE
Strange override behaviour with createAtStart=true

### ISSUE THREAD

**Issue by AVSilchenko** [InsertKoinIO/koin#1479]
Hello!
In our project we use `createAtStart=true` flag for faster service warm up. But after migration to 3.2.2 koin version override of modules starts work strange. Now overridden class instance is created too, and it might be unwanted behaviour in some cases. Example:
```
val moduleA = module {
    single(createdAtStart = true) {
        // instance creation may contain prod-specific logic, that is unwanted in testing where overridden module used
        SomeClassA()
    } bind SomeClassInterface::class
}

// should override moduleA:
val moduleB = module {
    single(createdAtStart = true) { // in previous koin version there was override=true here
        SomeClassB()
    } bind SomeClassInterface::class
}

install(Koin) {
    modules(moduleA + moduleB)
}
```
Override works properly, `get<SomeClassInterface>()` returns `SomeClassB` instance, but both `SomeClassA` and `SomeClassB` instances created

**To Reproduce**
code for reproduce problem is here https://github.com/AVSilchenko/kotlin-koin-beans-override

**Expected behaviour**
Creation only of overriding class instance

**Koin project used and used version:**
`koin-core version 3.2.2`

**Comment by arnaudgiuliani** [comment]
Hello,

Koin 3.2 (or perhaps before) introduced relaxed override rules.  The thing here is that your config exposes this: SomeClassA definition, and binding to SomeClassInterface. And a second definition with SomeClassB, with a bind to SomeClassInterface.

The type SomeClassInterface is indeed overridden, but not the original definition. 

To do such behavior, I would recommend writing:

```kotlin
val moduleA = module {
    single<SomeClassInterface>(createdAtStart = true) {
        SomeClassA()
    }
}

val moduleB = module {
    single<SomeClassInterface>(createdAtStart = true) {
        SomeClassB()
    }
}
```

**Comment by AVSilchenko** [comment]
Hello, @arnaudgiuliani !
Did you meant that original definition will fix problem with unwanted SomeClassA initialization?
If so - it is not. Tried to change my example code this way:
<img width="1414" alt="image" src="https://user-images.githubusercontent.com/15384075/211486711-cf4f3040-d10b-476d-967a-2ab2a39e6da3.png">

**Comment by arnaudgiuliani** [comment]
Yes I see, Both instances are still created 🤔

**Comment by arnaudgiuliani** [comment]
It'sa problem of indexing the right factories. 

will be released in koin-core 3.3.3
